Sources/Workspace/PinsStore.swift

```diff
@@ -61,27 +81,45 @@ public struct PinsStore {
     public var pins: AnySequence<Pin> {
         return AnySequence<Pin>(pinsMap.values)
     }
+    
+    /// Most recent error, if any.
+    /// FIXME: This needs to be an array of diagnostics, or perhaps a structured
+    /// log of what-all went on while trying to instantiate the PinStore.
+    public var error: Error?
+    
+    /// Returns true if and only iff there are any errors; when this property is
+    /// true, mutating PinStore operations are disallowed and will throw errors.
+    /// FIXME: We should have a protocol for "things that can hold errors", and
+    /// then the `hasErrors` property should be a convenience implementation so
+    /// it doesn't need to be implemented in every adopter of the protocol.
+    public var hasError: Bool { return error != nil }
```

Comment:
Ok.  I think a more interactive discussion is a good idea to settle on the overall approach.

I don't have strong feelings about PinStore itself — I started there because it was a leaf type and it seemed like a useful simple case for the broader approach, under the (perhaps mistaken) assumption that the same basic principle applied to most of the objects under Workspace.  Also, I couldn't start directly with Workspace, as explained in the description of the PR.

The one that's absolutely vital is Workspace; without the Workspace, there can be no package graph, and so a malformed JSON file down in the PinStore (for example) can prevent any package from being loaded at all.